### DISCUSSION TITLE
Deprecation of @Select decorator and proposed alternative

### DISCUSSION THREAD

**Discussion by markwhitfeld** [ngxs/store#1923] [RFC & Proposals]
## Background
There are issues with the decorator approach used with `@Select`:
- It lacks the ability to provide type safety or type inference
  - This is a typescript limitation that won't be fixed
- It uses a static global in order to gain access to the store
  - This causes issues with MFE apps and with server-side rendering in some situations.
  - See related issue: #1854
  - Many attempts at alternate approaches have had issues in other contexts and have eventually been reverted

We are looking to deprecate the `@Select` decorator in future versions of NGXS.

## Improved alternative

Just for a bit of a preview of what we are looking at, here is a prototype I made last year:
https://stackblitz.com/edit/ngxs-select-dispatch-utils-v0-bmecrf?file=src%2Fapp%2Fapp.component.ts

It is simpler, gives type inference and type safety and uses safe Angular mechanisms to get access to the `store`.

This example is in Angular 11, so there may be a minor tweak around the `ɵɵdirectiveInject` function based on the Angular version you are using.
In newer versions, there is an official `inject` function.
I have had some customers drop this utility in their codebase already, and are loving it!
Once we get more community feedback on this new approach and are confident that Angular has stabilised the `inject` api, then we will add some of these utils to the library.

We are pretty certain that we would be adding the `select` utility that you see here, potentially the `dispatch` utility (depending on how much feedback we get on the API for it), but most likely not the `selectSnapshot` utility (because it is not particularly performant in this current form), unless a better alternative is proposed.

**Comment by laurentsd** [comment]
In my team we have replaced all `@Select` with `store.select` since we upgraded to Angular 14 and are fine with it.

**Comment by joaqcid** [comment]
i still have some `@Select`s in my base code, but im ok with moving forward with this change
